### PR TITLE
Bring StreamWriter/Reader to CoreFX, add path constructors

### PR DESCRIPTION
**Pull request by ianhays**
- Exposes StreamWriter and StreamReader from CoreFX instead of mscorlib
- Adds the string constructors that take a path to a file to read/write/append
  - Because of the FileStream dependency, Reflection is used in a separate internal helper class to attain access to File.Open.
- Adds tests for the string constructors
- Note: StreamReader.ReadAsyncInternal isn't overriding the TextReader parent when building for netstandard1.7 because TextReader is still contained within mscorlib and it is in an internal method.

replaces https://github.com/dotnet/corefx/pull/12101
resolves https://github.com/dotnet/corefx/issues/12124

@stephentoub

### CODE REVIEW

File: src/System.IO/src/Resources/Strings.netcore50aot.resx [hunk 1]
```diff
@@ -225,4 +225,7 @@
   <data name="UnauthorizedAccess_MemStreamBuffer" xml:space="preserve">
     <value>MemoryStream's internal buffer cannot be accessed.</value>
   </data>
+  <data name="Argument_EmptyPath" xml:space="preserve">
+    <value>Empty path name is not le</value>
```
Comment (justinvp) [inline src/System.IO/src/Resources/Strings.netcore50aot.resx]: Is this string cut-off at the end?

File: src/System.IO/src/Resources/Strings.resx [hunk 2]
```diff
@@ -146,5 +146,26 @@
   </data>
   <data name="ObjectDisposed_StreamClosed" xml:space="preserve">
     <value>Cannot access a closed Stream.</value>
+  </data>  
+  <data name="InvalidOperation_AsyncIOInProgress" xml:space="preserve">
+    <value>The stream is currently in use by a previous operation on the stream.</value>
+  </data>
+  <data name="Argument_EmptyPath" xml:space="preserve">
+    <value>Empty path name is not le</value>
```
Comment (justinvp) [inline src/System.IO/src/Resources/Strings.resx]: Is this string cut-off at the end?

File: src/System.IO/src/Resources/Strings.netcore50aot.resx (hunk 1 above)
Comment (ianhays) [inline src/System.IO/src/Resources/Strings.netcore50aot.resx]: It is, thanks. Fixed.

File: src/System.IO/src/Resources/Strings.resx (hunk 2 above)
Comment (ianhays) [inline src/System.IO/src/Resources/Strings.resx]: It is, thanks. Fixed.

File: src/System.IO/src/System/IO/FileStreamHelpers.cs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    // permanent solution would be to merge the IO and FileSystem assemblies.
+    internal class FileStreamHelpers
+    {
+        private const int FileMode_Create = 2;
+        private const int FileMode_Open = 3;
+        private const int FileMode_Append = 6;
+        private const int FileAccess_Read = 1;
+        private const int FileAccess_Write = 2;
+
+        private static MethodInfo _fileOpen;
```
Comment (justinvp) [inline src/System.IO/src/System/IO/FileStreamHelpers.cs]: Nit: `s_fileOpen`

File: src/System.IO/src/System/IO/StreamReader.cs [hunk 4]
```diff
@@ -155,6 +155,48 @@ public StreamReader(Stream stream, Encoding encoding, bool detectEncodingFromByt
...
+            : this(path, encoding, true, DefaultBufferSize)
+        {
+        }
+
+        public StreamReader(string path, Encoding encoding, bool detectEncodingFromByteOrderMarks)
+            : this(path, encoding, detectEncodingFromByteOrderMarks, DefaultBufferSize)
+        {
+        }
+
+        [System.Security.SecuritySafeCritical]
```
Comment (justinvp) [inline src/System.IO/src/System/IO/StreamReader.cs]: Are the security attributes necessary?

File: src/System.IO/src/System/IO/StreamReader.cs [hunk 5]
```diff
@@ -155,6 +155,48 @@ public StreamReader(Stream stream, Encoding encoding, bool detectEncodingFromByt
...
+        public StreamReader(string path, Encoding encoding, bool detectEncodingFromByteOrderMarks, int bufferSize)
+            : this(path, encoding, detectEncodingFromByteOrderMarks, bufferSize, true)
+        {
+        }
+
+        [System.Security.SecurityCritical]
+        internal StreamReader(string path, Encoding encoding, bool detectEncodingFromByteOrderMarks, int bufferSize, bool checkHost)
```
Comment (justinvp) [inline src/System.IO/src/System/IO/StreamReader.cs]: Is `checkHost` even used here? If not, could this constructor be folded into the previous one?

File: src/System.IO/src/System/IO/StreamReader.cs [hunk 6]
```diff
@@ -155,6 +155,48 @@ public StreamReader(Stream stream, Encoding encoding, bool detectEncodingFromByt
...
+                throw new ArgumentNullException(nameof(path));
+            if (encoding == null)
+                throw new ArgumentNullException(nameof(encoding));
+            if (path.Length == 0)
+                throw new ArgumentException(SR.Argument_EmptyPath);
+            if (bufferSize <= 0)
+                throw new ArgumentOutOfRangeException(nameof(bufferSize), SR.ArgumentOutOfRange_NeedPosNum);
+
+            Stream stream = FileStreamHelpers.CreateFileStream(path, write: false, append: false);
+            Init(stream, encoding, detectEncodingFromByteOrderMarks, bufferSize, false);
```
Comment (justinvp) [inline src/System.IO/src/System/IO/StreamReader.cs]: Nit: `..., leaveOpen: false);`

File: src/System.IO/src/System/IO/FileStreamHelpers.cs [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+// See the LICENSE file in the project root for more information.
+
+using System.Reflection;
+
+namespace System.IO
+{
+    // Use reflection to access the implementation of FileStream in System.IO.FileSystem.dll. While far from ideal,
+    // we do this to avoid having to directly reference FileSystem as doing so introduces a dependency cycle. A more
+    // permanent solution would be to merge the IO and FileSystem assemblies.
+    internal class FileStreamHelpers
```
Comment (justinvp) [inline src/System.IO/src/System/IO/FileStreamHelpers.cs]: Nit: Should this be a static class?

Comment (ianhays) [inline src/System.IO/src/System/IO/FileStreamHelpers.cs]: yes.

File: src/System.IO/src/System/IO/StreamReader.cs (hunk 4 above)
Comment (ianhays) [inline src/System.IO/src/System/IO/StreamReader.cs]: I assumed so since they're included on the [mscorlib versions of these functions](https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/IO/StreamReader.cs#L191).

File: src/System.IO/src/System/IO/StreamReader.cs (hunk 6 above)
Comment (ianhays) [inline src/System.IO/src/System/IO/StreamReader.cs]: added.

File: src/System.IO/src/System.IO.csproj [hunk 8]
```diff
@@ -9,6 +9,7 @@
     <IsPartialFacadeAssembly>true</IsPartialFacadeAssembly>
     <NuGetTargetMoniker Condition="'$(TargetGroup)' == ''">.NETStandard,Version=v1.7</NuGetTargetMoniker>
     <PackageTargetFramework Condition="'$(TargetGroup)' == ''">netstandard1.7;uap10.1</PackageTargetFramework>
+    <NoWarn Condition="'$(TargetGroup)' != 'net462' and '$(TargetGroup)' != 'net463'">CS0436</NoWarn>
```
Comment (ianhays) [inline src/System.IO/src/System.IO.csproj]: cc: @weshaggard this is the ignore we chatted about.

File: src/System.IO/src/System/IO/FileStreamHelpers.cs [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+        public static Stream CreateFileStream(string path, bool write, bool append)
+        {
+            if (s_fileOpen == null)
+            {
+                s_fileOpen = GetFileOpenFunction();
+            }
+            int filemode = !write ? FileMode_Open : append ? FileMode_Append : FileMode_Create;
+            int fileaccess = write ? FileAccess_Write : FileAccess_Read;
+            object[] parameters = new object[] { path, filemode, fileaccess };
+            return s_fileOpen?.Invoke(null, parameters) as Stream;
```
Comment (stephentoub) [inline src/System.IO/src/System/IO/FileStreamHelpers.cs]: You'd need to measure, but I think it'll be measurably faster to create a delegate for the MethodInfo in GetFileOpenFunction using MethodInfo.CreateDelegate, store that delegate in the static field, and then just invoke that delegate here as if you were calling a function (if nothing else, it'll avoid needing to manually allocate the array).  I realize that's not how it's done in the Process-related code you copied; we should probably change that there, too.  I also realize this is only meant to be a temporary hack, but just in case :)

File: src/System.IO/src/System/IO/FileStreamHelpers.cs [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+        }
+
+        private static MethodInfo GetFileOpenFunction()
+        {
+            Type fileSystemType = Type.GetType("System.IO.File, System.IO.FileSystem, Version=4.0.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a", throwOnError: false);
+            var methodInfos = fileSystemType?.GetTypeInfo().GetDeclaredMethods("Open");
+            if (methodInfos != null)
+            {
+                foreach (var methodInfo in methodInfos)
```
Comment (stephentoub) [inline src/System.IO/src/System/IO/FileStreamHelpers.cs]: Nit: var => MethodInfo

File: src/System.IO/src/System/IO/FileStreamHelpers.cs [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+            int filemode = !write ? FileMode_Open : append ? FileMode_Append : FileMode_Create;
+            int fileaccess = write ? FileAccess_Write : FileAccess_Read;
+            object[] parameters = new object[] { path, filemode, fileaccess };
+            return s_fileOpen?.Invoke(null, parameters) as Stream;
+        }
+
+        private static MethodInfo GetFileOpenFunction()
+        {
+            Type fileSystemType = Type.GetType("System.IO.File, System.IO.FileSystem, Version=4.0.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a", throwOnError: false);
+            var methodInfos = fileSystemType?.GetTypeInfo().GetDeclaredMethods("Open");
```
Comment (stephentoub) [inline src/System.IO/src/System/IO/FileStreamHelpers.cs]: Could this be done using the GetMethods overload that accepts a Type[], avoiding the need to manually loop through the overloads?

File: src/System.IO/src/System/IO/StreamReader.cs [hunk 12]
```diff
@@ -155,6 +155,41 @@ public StreamReader(Stream stream, Encoding encoding, bool detectEncodingFromByt
...
+            if (path.Length == 0)
+                throw new ArgumentException(SR.Argument_EmptyPath);
+            if (bufferSize <= 0)
+                throw new ArgumentOutOfRangeException(nameof(bufferSize), SR.ArgumentOutOfRange_NeedPosNum);
+
+            Stream stream = FileStreamHelpers.CreateFileStream(path, write: false, append: false);
+            Init(stream, encoding, detectEncodingFromByteOrderMarks, bufferSize, leaveOpen: false);
+        }
+
         private void Init(Stream stream, Encoding encoding, bool detectEncodingFromByteOrderMarks, int bufferSize, bool leaveOpen)
```
Comment (stephentoub) [inline src/System.IO/src/System/IO/StreamReader.cs]: Have you diff'd the StreamReader/Writer code in corefx against the code in coreclr?  It'd be good to verify they haven't diverged due to changes made in coreclr that didn't make it back to corefx.

File: src/System.IO/src/System/IO/StreamReader.cs [hunk 13]
```diff
@@ -962,8 +997,11 @@ public override Task<int> ReadAsync(char[] buffer, int index, int count)
 
             return task;
         }
-
+#if uap101aot
```
Comment (stephentoub) [inline src/System.IO/src/System/IO/StreamReader.cs]: Is there an issue# associated with this?  Could you include a TODO with it for this to be removed?

File: src/System.IO/tests/StreamReader/StreamReader.StringCtorTests.cs [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+        [Theory]
+        [InlineData(true)]
+        [InlineData(false)]
+        public static void ReadToEnd_detectEncodingFromByteOrderMarks(bool detectEncodingFromByteOrderMarks)
+        {
+            string testfile = Path.GetTempFileName();
+            try
+            {
+                File.WriteAllBytes(testfile, new byte[] { 65, 66, 67, 68 });
+                using (var sr2 = new StreamReader(testfile, detectEncodingFromByteOrderMarks))
```
Comment (stephentoub) [inline src/System.IO/tests/StreamReader/StreamReader.StringCtorTests.cs]: Do we have tests to exercise all of the ctors that were just added?

File: src/System.IO/src/System/IO/StreamReader.cs (hunk 13 above)
Comment (jkotas) [inline src/System.IO/src/System/IO/StreamReader.cs]: Is there any problem with moving the TextReader/Writer out of mscorlib?

File: src/System.IO/src/System/IO/FileStreamHelpers.cs (hunk 11 above)
Comment (ianhays) [inline src/System.IO/src/System/IO/FileStreamHelpers.cs]: The problem is that we don't have access to FileAccess or FileMode, so unless we reflect those as well then we can't get `Open` that way.

Comment (ianhays) [inline src/System.IO/src/System/IO/FileStreamHelpers.cs]: >I guess System.IO can't depend on System.IO.FileSystem.Primitives

I get a great big helping of errors (e.g. System.Object is not defined or imported) when I attempt that.

>We should really just combine all of these (after your change).

Absolutely.

File: src/System.IO/src/System/IO/FileStreamHelpers.cs (hunk 9 above)
Comment (ianhays) [inline src/System.IO/src/System/IO/FileStreamHelpers.cs]: I wasn't aware of methodInfo.CreateDelegate. I prefer that stylistically to what I have.

File: src/System.IO/src/System/IO/StreamReader.cs [hunk 15]
```diff
@@ -480,6 +521,14 @@ private void DetectEncoding()
                 CompressBuffer(3);
                 changedEncoding = true;
             }
+            else if (_byteLen >= 4 && _byteBuffer[0] == 0 && _byteBuffer[1] == 0 &&
+                _byteBuffer[2] == 0xFE && _byteBuffer[3] == 0xFF)
+            {
+                // Big Endian UTF32
+                _encoding = new UTF32Encoding(true, true);
```
Comment (justinvp) [inline src/System.IO/src/System/IO/StreamReader.cs]: Nit: `new UTF32Encoding(bigEndian: true, byteOrderMark: true);`

Alternatively, the following would allow caching/reusing the `UTF32Encoding` instance, based on the current code in coreclr:

```c#
const int CodePageUTF32BE = 12001;
_encoding = Encoding.GetEncoding(CodePageUTF32BE);
```

Comment (ianhays) [inline src/System.IO/src/System/IO/StreamReader.cs]: Thanks for the quick feedback on the update! I pushed another with the parameter names. The GetEncoding use looks good, but I'd rather we do that in a different PR and keep this one  as minimum a set of porting changes as possible.
